kafka/proxy.go

```diff
@@ -109,58 +146,70 @@ func (h *produceRequestHandler) Handle(requestKeyVersion *kafkaprotocol.RequestK
 		return shouldReply, nil
 	}
 
-	msgs := h.extractMessages(req)
+	msgs, err := h.extractMessages(req)
+	if err != nil {
+		logrus.WithError(err).Error("error extracting messages")
+		return shouldReply, nil
+	}
+
 	if len(msgs) == 0 {
 		logrus.Error("The produce request has no messages")
 		return
 	}
 
-	for _, m := range msgs {
-		if _, err := h.chain(m); err != nil {
-			logrus.WithError(err).Error("error handling message")
-			return shouldReply, nil
-		}
+	if err := h.publisher.Publish(messagesChannelName, msgs...); err != nil {
+		logrus.WithError(err).Error("error handling message")
+		return shouldReply, nil
 	}
 
 	return shouldReply, nil
 }
 
-func (h *produceRequestHandler) extractMessages(req sarama.ProduceRequest) []*message.Message {
-	var msgs []*message.Message
-	for topic, r := range req.Records {
-		for _, s := range r {
+func (h *produceRequestHandler) extractMessages(req sarama.ProduceRequest) ([]*watermillmessage.Message, error) {
+	marshaler := watermillkafka.DefaultMarshaler{}
```

Comment:
Does `marshaller` (btw. you missed about second `l`) have to be created for each function call? Maybe it would be better to create singleton outside of function.

Reply:
Go uses the term `Marshaler` and `Unmashaler` with just one `l`, both for referring to the interface and for the verb `Marshaling` or `Unmarshaling`. E.g. https://pkg.go.dev/encoding/json#Marshaler

👍  to move it into a var so it is only instantiated once. Btw, this is now done!